File: Data Structures & Algorithms/house-robber-ii/submission-3.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        if(nums.size() == 1){
            return nums[0];
        } else if (nums.size() == 2){
            return max(nums[0], nums[1]);
        }
        int n = nums.size();
        vector<int> rob(n, 0);
        rob[0] = nums[0], rob[1] = max(nums[0], nums[1]);
        int ans = 0;
        for(int i = 2; i < n-1; i++){
            rob[i] = max(rob[i-1], nums[i] + rob[i-2]);
            ans = max(ans, rob[i]);
        }

        rob[0] = 0, rob[1] = nums[1];
        for(int i = 2; i < n; i++){
            rob[i] = max(rob[i-1], nums[i] + rob[i-2]);
            ans = max(ans, rob[i]);
        }
        
        return ans;

    }
};

File: Data Structures & Algorithms/top-k-elements-in-list/submission-3.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> count = new HashMap<>();

        for (int num : nums) {
            count.put(num, count.getOrDefault(num, 0) + 1);
        }

        List<Integer>[] buckets = new ArrayList[nums.length + 1];
        for (int i = 0; i <= nums.length; i++) {
            buckets[i] = new ArrayList<>();
        }

        for (Map.Entry<Integer, Integer> entry : count.entrySet()) {
            buckets[entry.getValue()].add(entry.getKey());
        }

        int[] ans = new int[k];
        int i = 0;
        for (int j = nums.length; j >= 0 && i < k; j--) {
            for (int l = 0; l < buckets[j].size(); l++) {
                if (i == k) {
                    return ans;
                }
                ans[i++] = buckets[j].get(l);
            }
        }

        return ans;
    }
}